src/main/java/io/nats/client/impl/Headers.java

```diff
@@ -0,0 +1,217 @@
+// Copyright 2015-2018 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io.nats.client.impl;
+
+import java.util.*;
+
+public class Headers {
+	private static final String KEY_CANNOT_BE_EMPTY_OR_NULL = "Header key cannot be null.";
+	private static final String KEY_INVALID_CHARACTER = "Header key has invalid character: ";
+	private static final String VALUE_INVALID_CHARACTERS = "Header value has invalid character: ";
+
+	private final Map<String, List<String>> headerMap = new HashMap<>();
+	private final Map<String, String> valueCsvCache = new HashMap<>();
```

Comment:
It should be possible to use a `ByteBuffer` based hashmap here like I've done in #356 which makes message serialization simpler in general as Java strings(which are UTF-16 based) do not properly map to the wire protocol(which is ASCII/UTF-8 based).